jonprl-mode.el

```diff
@@ -200,6 +205,89 @@ This list is constructed from JonPRL's output.")
+                                       nil)))
+                      (cond
+                       (matching matching)
+                       ;; The parent of ~ is nil and the parent of / is itself.
+                       ;; Thus the terminating condition for not finding the file
+                       ;; accounts for both.
+                       ((or (null parent) (equal parent (directory-file-name parent))) nil) ; Not found
+                       (t (find-file-r (directory-file-name parent))))))) ; Continue
+    (let* ((file (buffer-file-name (current-buffer)))
+           (dir (if file (file-name-directory file) default-directory)))
+      (when dir
+        (cl-remove-if #'(lambda (f)
+                          (and (not allow-hidden)
+                               (string-prefix-p "." (file-name-nondirectory f))))
+                      (find-file-r dir))))))
+
+;; The following code is cribbed from idris-ipkg-mode
+(defun jonprl-try-directory-files (directory &optional full match nosort)
+  "Call `directory-files' with arguments DIRECTORY, FULL, MATCH,
+and NOSORT, but return the empty list on failure instead of
+throwing an error.
+See the docstring for `directory-files' for the meaning of the
+arguments."
+  ;; This wrapper is useful because some users can't read all the
+  ;; directories above the current working directory. In particular,
+  ;; /home is sometimes not readable.
+  (condition-case nil
+      (directory-files directory full match nosort)
+    (error nil)))
+
+
+(defun jonprl-command-args (should-print)
+  (let* ((filename (buffer-file-name))
+         (file (file-name-nondirectory filename))
+         (config-file jonprl-configuration-file))
+    (concat
+     (if (not should-print) "--check " "")
+       (if (or (not config-file) (equal config-file 'shut-up)) file config-file))))
+
+(defun jonprl-forget-configuration ()
+  "Forgets the configuration file"
+  (interactive)
+  (setq jonprl-configuration-file nil)
+  (message "Configuration file unset"))
+
+(defun jonprl-read-configuration-file ()
+  (let ((cfg-prompt "Set configuration file:")
+        (cfg-default (jonprl-find-file-upwards "cfg")))
+    (expand-file-name
+      (if cfg-default
+        (read-file-name cfg-prompt
+                        (file-name-directory (car cfg-default))
+                        (car cfg-default)
+                        t
+                        (file-name-nondirectory (car cfg-default)))
+        (read-file-name cfg-prompt nil nil nil t)))))
+
+(defun jonprl-set-configuration-file (filename)
+  "Choose a configuration file for your JonPRL development"
+  (interactive (list (jonprl-read-configuration-file)))
+  (setq jonprl-configuration-file filename)
+  (message (concat "Configuration file set to " filename)))
+
+(defun jonprl-initialize-configuration ()
+  (interactive)
+  (unless jonprl-configuration-file
+    (if (y-or-n-p "Do you want to choose a .cfg file?")
```

Comment:
Why not something like

```
(let ((cfg (read-file-name "Set configuration (leave blank for none): " ...)))
  (if (= (length cfg) 0) nil cfg))
```
to read it? This works OK for `idris-set-idris-load-packages` in `idris-mode`. `C-g` doesn't work because that causes the whole function to bail - you could also try wrapping `read-file-name` in `ignore-errors` to catch the `C-g`.